File: MaskInfo/app/src/main/java/org/techtown/maskinfo/MainViewModel.java
package org.techtown.maskinfo;

import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import org.techtown.maskinfo.model.Store;
import org.techtown.maskinfo.model.StoreInfo;
import org.techtown.maskinfo.repository.MaskService;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

public class MainViewModel extends ViewModel {

    private static final String TAG = MainViewModel.class.getSimpleName();
    public MutableLiveData<List<Store>> itemLiveData = new MutableLiveData<>();

    private Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(MaskService.BASE_URL)
            .addConverterFactory(MoshiConverterFactory.create())
            .build();

    private MaskService service = retrofit.create(MaskService.class);

    private Call<StoreInfo> storeInfoCall = service.fetchStoreInfo();

    public MainViewModel(){
        fetchStoreInfo();
    }

    public void fetchStoreInfo(){
        //비동기 처리.
        storeInfoCall.clone().enqueue(new Callback<StoreInfo>() {
            @Override
            public void onResponse(Call<StoreInfo> call, Response<StoreInfo> response) {
                Log.d(TAG,"onResponse:: refresh");
                List<Store> items = response.body().getStores()
                        .stream()
                        .filter(item -> item.getRemainStat() != null)
                        .collect(Collectors.toList());

                //백그라운드 에선 postValue.
                itemLiveData.postValue(items);
            }

            @Override
            public void onFailure(Call<StoreInfo> call, Throwable t) {
                Log.e(TAG,"onFailure:",t);
                //빈거 세팅해주기.
                itemLiveData.postValue(Collections.emptyList());
            }
        });
    }
}

File: Counter/app/src/main/java/org/techtown/counter/MainViewModel.kt
package org.techtown.counter

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel

class MainViewModel(private val handle: SavedStateHandle) : ViewModel() {
    private var count = handle.get<Int>("count") ?: 0 //saveInstance값 가져옴.(강종되었을때)
        set(value) {//위에 count값이 바뀌게 되면은 value값으로 들어오게 될거야.
            field = value //해당값을 확정짓겟다하면 field에 넣어둠.
            countLiveData.value = value
            handle.set("count", value) //saveInstance값 세팅해줌.
        }

    val countLiveData = MutableLiveData<Int>()

    fun increaseCount(){
        count++
    }

    fun decreaseCount(){
        count--
    }
}